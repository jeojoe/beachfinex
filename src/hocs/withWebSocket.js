import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const ActionRow = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 10px 0 0;
`;

function withWebSocket(Component) {
  return class WithWebSocket extends React.Component {
    state = {
      subscribed: false,
      subscribing: true,
      actionRow: null,
    }

    subscribe = ({ newOpenMsg, newOnMessage, newActionRow } = {}) => {
      // New settings
      if (newOpenMsg) this.openMsg = newOpenMsg;
      if (newOnMessage) this.onMessage = newOnMessage;
      if (newActionRow) this.setState({ actions: newActionRow });
      // New WS instance
      this.ws = new WebSocket('wss://api.bitfinex.com/ws/2');
      this.ws.onopen = this.onOpen;
      this.ws.onmessage = this.onMessage;
      this.ws.onerror = this.onError;
      this.setState({ subscribing: true });
    }

    onOpen = () => this.ws.send(JSON.stringify(this.openMsg));

    onError = err => this.unsubscribe(err);

    unsubscribe = (err) => {
      this.ws.close();
      this.ws = null;
      this.setState({ subscribed: false });
      if (err) {
        // Has error, not user's action => subscribe again
        console.error(err);
        this.subscribe();
      }
    }

    toggle = () => {
      if (this.ws) {
        this.unsubscribe();
      } else {
        this.subscribe();
      }
    }

    setSubscribed = () => this.setState({
      subscribed: true,
      subscribing: false,
    });

    renderActions() {
      const { subscribed, subscribing, actions } = this.state;
      return (
        <ActionRow>
          <div>
            <b>{subscribed ? 'REAL-TIME' : 'OFFLINE' }</b>
            {' '}
            {subscribing
              ? '(Subscribing..)'
              : (
                <button onClick={this.toggle} type="button">
                  {subscribed ? 'Unsubscribe' : 'Subscribe'}
                </button>
              )
            }
          </div>
          {actions}
        </ActionRow>
      );
    }

    render() {
      const { subscribed, subscribing } = this.state;

      return (
        <div>
          <Component
            {...this.props}
            ws={{
              subscribed,
              subscribing,
              setSubscribed: this.setSubscribed,
              subscribe: this.subscribe,
              unsubscribe: this.unsubscribe,
              toggle: this.toggle,
            }}
          />
          {this.renderActions()}
        </div>
      );
    }
  };
}

export const propTypesWS = {
  subscribed: PropTypes.bool.isRequired,
  subscribing: PropTypes.bool.isRequired,
  setSubscribed: PropTypes.func.isRequired,
  subscribe: PropTypes.func.isRequired,
  unsubscribe: PropTypes.func.isRequired,
  toggle: PropTypes.func.isRequired,
};

export default withWebSocket;
