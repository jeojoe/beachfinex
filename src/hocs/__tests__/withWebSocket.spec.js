import React from 'react';
import { shallow } from 'enzyme';

import withWebSocket from '../withWebSocket';

describe('withWebSocket()', () => {
  it('enhances input component correctly', () => {
    const cmp = () => <div />;
    const Enhanced = withWebSocket(cmp);
    expect(shallow(<Enhanced />)).toMatchSnapshot();
  });
});
