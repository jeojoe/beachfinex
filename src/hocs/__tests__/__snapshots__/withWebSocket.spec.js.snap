// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withWebSocket() enhances input component correctly 1`] = `
<div>
  <cmp
    ws={
      Object {
        "setMoreActions": [Function],
        "setSubscribed": [Function],
        "subscribe": [Function],
        "subscribed": false,
        "subscribing": true,
        "toggle": [Function],
        "unsubscribe": [Function],
      }
    }
  />
  <styled.div>
    <div>
      <b>
        OFFLINE
      </b>
       
      (Subscribing..)
    </div>
  </styled.div>
</div>
`;
